--- httpd-2.2.15/support/ab.c.aboverflow
+++ httpd-2.2.15/support/ab.c
@@ -1577,6 +1577,10 @@ static void test(void)
 
     stats = calloc(requests, sizeof(struct data));
 
+    if (!con || !stats) {
+        err("Cannot allocate memory for result statistics");
+    }
+
     if ((status = apr_pollset_create(&readbits, concurrency, cntxt, 0)) != APR_SUCCESS) {
         apr_err("apr_pollset_create failed", status);
     }
