--- httpd-2.2.15/modules/proxy/mod_proxy_ajp.c.ORIG	2011-12-13 14:58:16.815627078 +0100
+++ httpd-2.2.15/modules/proxy/mod_proxy_ajp.c	2011-12-13 14:59:39.122512315 +0100
@@ -492,30 +492,32 @@
                                                  conn->worker->flush_wait))
                                                  != APR_SUCCESS) &&
                                   APR_STATUS_IS_TIMEUP(rv))) {
                                 e = apr_bucket_flush_create(r->connection->bucket_alloc);
                                 APR_BRIGADE_INSERT_TAIL(output_brigade, e);
                             }
                             apr_brigade_length(output_brigade, 0, &bb_len);
                             if (bb_len != -1)
                                 conn->worker->s->read += bb_len;
                         }
-                        if (ap_pass_brigade(r->output_filters,
-                                            output_brigade) != APR_SUCCESS) {
-                            ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r,
-                                          "proxy: error processing body.%s",
-                                          r->connection->aborted ?
-                                          " Client aborted connection." : "");
-                            output_failed = 1;
-                        }
-                        data_sent = 1;
-                        apr_brigade_cleanup(output_brigade);
+			if (headers_sent) {
+				if (ap_pass_brigade(r->output_filters,
+						    output_brigade) != APR_SUCCESS) {
+				    ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r,
+						  "proxy: error processing body.%s",
+						  r->connection->aborted ?
+						  " Client aborted connection." : "");
+				    output_failed = 1;
+				}
+				data_sent = 1;
+				apr_brigade_cleanup(output_brigade);
+			}
                     }
                 }
                 else {
                     backend_failed = 1;
                 }
                 break;
             case CMD_AJP13_END_RESPONSE:
                 /* If we are overriding the errors, we must not send anything to
                  * the client, especially as the brigade already contains headers.
                  * So do nothing here, and it will be cleaned up below.
